File: android/app/src/main/kotlin/com/example/pixeldiary/ble/BlePeripheralManager.kt
package com.example.pixeldiary.ble

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothGattServer
import android.bluetooth.BluetoothGattServerCallback
import android.bluetooth.BluetoothGattService
import android.bluetooth.BluetoothManager
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.content.Context
import android.os.ParcelUuid
import android.util.Log
import java.nio.charset.StandardCharsets
import java.util.UUID

/// BLE Peripheral Manager
/// アドバタイズとGATTサーバーを管理
class BlePeripheralManager(private val context: Context) {
    companion object {
        private const val TAG = "BlePeripheralManager"

        // PixelDiary サービスUUID
        val SERVICE_UUID: UUID = UUID.fromString("4f4d4449-5843-4841-4e47-455f53455256")

        // Write Characteristic UUID
        val WRITE_CHAR_UUID: UUID = UUID.fromString("4f4d4449-5843-4841-4e47-455f575249")

        // Read Characteristic UUID (Notify)
        val READ_CHAR_UUID: UUID = UUID.fromString("4f4d4449-5843-4841-4e47-455f52454144")

        // Pairing Characteristic UUID
        val PAIRING_CHAR_UUID: UUID = UUID.fromString("4f4d4449-5843-4841-4e47-455f50414952")

        // アドバタイズプレフィックス
        private const val ADVERTISE_PREFIX = "PD_"

        // Client Characteristic Configuration Descriptor
        val CCCD_UUID: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
    }

    private val bluetoothManager: BluetoothManager =
        context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager

    private val bluetoothAdapter: BluetoothAdapter? = bluetoothManager.adapter
    private val advertiser: BluetoothLeAdvertiser? = bluetoothAdapter?.bluetoothLeAdvertiser

    private var gattServer: BluetoothGattServer? = null
    private var isAdvertising = false
    private var currentNickname: String? = null
    private var currentArtData: String? = null

    // コールバック
    var onDataReceived: ((data: String, device: BluetoothDevice) -> Unit)? = null
    var onDeviceConnected: ((device: BluetoothDevice) -> Unit)? = null
    var onDeviceDisconnected: ((device: BluetoothDevice) -> Unit)? = null
    var onError: ((message: String) -> Unit)? = null

    // 接続中のデバイス
    private val connectedDevices = mutableSetOf<BluetoothDevice>()

    /// アドバタイズ開始
    fun startAdvertising(nickname: String, artData: String): Boolean {
        if (advertiser == null) {
            Log.e(TAG, "Bluetooth LE Advertiser not available")
            onError?.invoke("アドバタイズ機能が利用できません")
            return false
        }

        if (isAdvertising) {
            Log.w(TAG, "Already advertising")
            return true
        }

        currentNickname = nickname
        currentArtData = artData

        // GATTサーバーをセットアップ
        if (!setupGattServer()) {
            return false
        }

        // アドバタイズ設定
        val settings = AdvertiseSettings.Builder()
            .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_BALANCED)
            .setConnectable(true)
            .setTimeout(0) // 無期限
            .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_MEDIUM)
            .build()

        // Bluetoothアダプタのローカル名を設定（アドバタイズに含めるため）
        val localName = "$ADVERTISE_PREFIX$nickname"
        try {
            bluetoothAdapter?.name = localName
            Log.d(TAG, "Bluetooth local name set to: $localName")
        } catch (e: SecurityException) {
            Log.w(TAG, "Cannot set Bluetooth name due to missing permission", e)
        }

        // アドバタイズデータ（サービスUUID）
        val data = AdvertiseData.Builder()
            .addServiceUuid(ParcelUuid(SERVICE_UUID))
            .setIncludeDeviceName(false)
            .setIncludeTxPowerLevel(false)
            .build()

        // スキャンレスポンスデータ（ニックネームをローカル名として含める）
        val scanResponse = AdvertiseData.Builder()
            .setIncludeDeviceName(true)  // ローカル名を含める
            .setIncludeTxPowerLevel(false)
            .build()

        // アドバタイズ開始
        try {
            advertiser.startAdvertising(settings, data, scanResponse, advertiseCallback)
            Log.i(TAG, "Advertising started with nickname: $nickname")
            return true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start advertising", e)
            onError?.invoke("アドバタイズ開始に失敗しました: ${e.message}")
            return false
        }
    }

    /// アドバタイズ停止
    fun stopAdvertising() {
        if (!isAdvertising) {
            return
        }

        try {
            advertiser?.stopAdvertising(advertiseCallback)
            isAdvertising = false
            Log.i(TAG, "Advertising stopped")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to stop advertising", e)
        }

        // GATTサーバークローズ
        closeGattServer()
    }

    /// GATTサーバーセットアップ
    private fun setupGattServer(): Boolean {
        try {
            gattServer = bluetoothManager.openGattServer(context, gattServerCallback)

            if (gattServer == null) {
                Log.e(TAG, "Failed to open GATT server")
                onError?.invoke("GATTサーバー起動に失敗しました")
                return false
            }

            // サービス作成
            val service = BluetoothGattService(
                SERVICE_UUID,
                BluetoothGattService.SERVICE_TYPE_PRIMARY
            )

            // Write Characteristic
            val writeChar = BluetoothGattCharacteristic(
                WRITE_CHAR_UUID,
                BluetoothGattCharacteristic.PROPERTY_WRITE or
                        BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
                BluetoothGattCharacteristic.PERMISSION_WRITE
            )
            service.addCharacteristic(writeChar)

            // Read Characteristic (Notify)
            val readChar = BluetoothGattCharacteristic(
                READ_CHAR_UUID,
                BluetoothGattCharacteristic.PROPERTY_READ or
                        BluetoothGattCharacteristic.PROPERTY_NOTIFY,
                BluetoothGattCharacteristic.PERMISSION_READ
            )

            // CCCD Descriptor（Notify有効化用）
            val cccdDescriptor = BluetoothGattDescriptor(
                CCCD_UUID,
                BluetoothGattDescriptor.PERMISSION_READ or
                        BluetoothGattDescriptor.PERMISSION_WRITE
            )
            readChar.addDescriptor(cccdDescriptor)
            service.addCharacteristic(readChar)

            // Pairing Characteristic（将来の拡張用）
            val pairingChar = BluetoothGattCharacteristic(
                PAIRING_CHAR_UUID,
                BluetoothGattCharacteristic.PROPERTY_WRITE or
                        BluetoothGattCharacteristic.PROPERTY_READ,
                BluetoothGattCharacteristic.PERMISSION_WRITE or
                        BluetoothGattCharacteristic.PERMISSION_READ
            )
            service.addCharacteristic(pairingChar)

            // サービス追加
            gattServer?.addService(service)

            Log.i(TAG, "GATT server setup completed")
            return true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to setup GATT server", e)
            onError?.invoke("GATTサーバーセットアップに失敗しました: ${e.message}")
            return false
        }
    }

    /// GATTサーバークローズ
    private fun closeGattServer() {
        try {
            // 接続中のデバイスを切断
            connectedDevices.forEach { device ->
                gattServer?.cancelConnection(device)
            }
            connectedDevices.clear()

            gattServer?.close()
            gattServer = null
            Log.i(TAG, "GATT server closed")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to close GATT server", e)
        }
    }

    /// アドバタイズコールバック
    private val advertiseCallback = object : AdvertiseCallback() {
        override fun onStartSuccess(settingsInEffect: AdvertiseSettings) {
            isAdvertising = true
            Log.i(TAG, "Advertise started successfully")
        }

        override fun onStartFailure(errorCode: Int) {
            isAdvertising = false
            val message = when (errorCode) {
                ADVERTISE_FAILED_DATA_TOO_LARGE -> "データサイズが大きすぎます"
                ADVERTISE_FAILED_TOO_MANY_ADVERTISERS -> "アドバタイザーが多すぎます"
                ADVERTISE_FAILED_ALREADY_STARTED -> "既にアドバタイズ中です"
                ADVERTISE_FAILED_INTERNAL_ERROR -> "内部エラーが発生しました"
                ADVERTISE_FAILED_FEATURE_UNSUPPORTED -> "機能がサポートされていません"
                else -> "不明なエラー: $errorCode"
            }
            Log.e(TAG, "Advertise failed: $message")
            onError?.invoke("アドバタイズ失敗: $message")
        }
    }

    /// GATTサーバーコールバック
    private val gattServerCallback = object : BluetoothGattServerCallback() {
        override fun onConnectionStateChange(
            device: BluetoothDevice,
            status: Int,
            newState: Int
        ) {
            when (newState) {
                BluetoothGatt.STATE_CONNECTED -> {
                    connectedDevices.add(device)
                    Log.i(TAG, "Device connected: ${device.address}")
                    onDeviceConnected?.invoke(device)
                }
                BluetoothGatt.STATE_DISCONNECTED -> {
                    connectedDevices.remove(device)
                    Log.i(TAG, "Device disconnected: ${device.address}")
                    onDeviceDisconnected?.invoke(device)
                }
            }
        }

        override fun onCharacteristicWriteRequest(
            device: BluetoothDevice,
            requestId: Int,
            characteristic: BluetoothGattCharacteristic,
            preparedWrite: Boolean,
            responseNeeded: Boolean,
            offset: Int,
            value: ByteArray
        ) {
            when (characteristic.uuid) {
                WRITE_CHAR_UUID -> {
                    // データ受信
                    val receivedData = String(value, StandardCharsets.UTF_8)
                    Log.d(TAG, "Received data from ${device.address}: ${receivedData.take(50)}...")

                    // 受信データをコールバック
                    onDataReceived?.invoke(receivedData, device)

                    // 自分のデータを送信
                    sendDataToDevice(device)

                    // レスポンス送信
                    if (responseNeeded) {
                        gattServer?.sendResponse(
                            device,
                            requestId,
                            BluetoothGatt.GATT_SUCCESS,
                            offset,
                            value
                        )
                    }
                }
                else -> {
                    Log.w(TAG, "Unknown characteristic write: ${characteristic.uuid}")
                    if (responseNeeded) {
                        gattServer?.sendResponse(
                            device,
                            requestId,
                            BluetoothGatt.GATT_FAILURE,
                            offset,
                            null
                        )
                    }
                }
            }
        }

        override fun onCharacteristicReadRequest(
            device: BluetoothDevice,
            requestId: Int,
            offset: Int,
            characteristic: BluetoothGattCharacteristic
        ) {
            when (characteristic.uuid) {
                READ_CHAR_UUID -> {
                    // 自分のデータを返す
                    val data = currentArtData?.toByteArray(StandardCharsets.UTF_8) ?: byteArrayOf()
                    Log.d(TAG, "Read request from ${device.address}, sending ${data.size} bytes")

                    gattServer?.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        offset,
                        data.copyOfRange(offset, data.size)
                    )
                }
                else -> {
                    Log.w(TAG, "Unknown characteristic read: ${characteristic.uuid}")
                    gattServer?.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_FAILURE,
                        offset,
                        null
                    )
                }
            }
        }

        override fun onDescriptorWriteRequest(
            device: BluetoothDevice,
            requestId: Int,
            descriptor: BluetoothGattDescriptor,
            preparedWrite: Boolean,
            responseNeeded: Boolean,
            offset: Int,
            value: ByteArray
        ) {
            if (descriptor.uuid == CCCD_UUID) {
                // Notify有効化
                Log.d(TAG, "CCCD write request from ${device.address}")

                if (responseNeeded) {
                    gattServer?.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_SUCCESS,
                        offset,
                        value
                    )
                }
            } else {
                Log.w(TAG, "Unknown descriptor write: ${descriptor.uuid}")
                if (responseNeeded) {
                    gattServer?.sendResponse(
                        device,
                        requestId,
                        BluetoothGatt.GATT_FAILURE,
                        offset,
                        null
                    )
                }
            }
        }
    }

    /// デバイスにデータ送信（Notify）
    private fun sendDataToDevice(device: BluetoothDevice) {
        val data = currentArtData?.toByteArray(StandardCharsets.UTF_8) ?: return

        val service = gattServer?.getService(SERVICE_UUID) ?: return
        val characteristic = service.getCharacteristic(READ_CHAR_UUID) ?: return

        characteristic.value = data

        try {
            gattServer?.notifyCharacteristicChanged(device, characteristic, false)
            Log.d(TAG, "Notified data to ${device.address}: ${data.size} bytes")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to notify data", e)
        }
    }

    /// クリーンアップ
    fun cleanup() {
        stopAdvertising()
        closeGattServer()
        connectedDevices.clear()
        currentNickname = null
        currentArtData = null
    }

    /// 状態確認
    fun isAdvertising(): Boolean = isAdvertising

    /// 接続デバイス数
    fun getConnectedDeviceCount(): Int = connectedDevices.size
}

File: ios/Runner/Ble/BlePeripheralManager.swift
import Foundation
import CoreBluetooth

/// BLE Peripheral Manager
/// アドバタイズとGATTサーバーを管理
class BlePeripheralManager: NSObject {
    // MARK: - UUIDs

    // PixelDiary サービスUUID
    static let serviceUUID = CBUUID(string: "4F4D4449-5843-4841-4E47-455F53455256")

    // Write Characteristic UUID
    static let writeCharUUID = CBUUID(string: "4F4D4449-5843-4841-4E47-455F575249")

    // Read Characteristic UUID (Notify)
    static let readCharUUID = CBUUID(string: "4F4D4449-5843-4841-4E47-455F52454144")

    // Pairing Characteristic UUID
    static let pairingCharUUID = CBUUID(string: "4F4D4449-5843-4841-4E47-455F50414952")

    // アドバタイズプレフィックス
    private static let advertisePrefix = "PD_"

    // MARK: - Properties

    private var peripheralManager: CBPeripheralManager!
    private var service: CBMutableService?
    private var writeCharacteristic: CBMutableCharacteristic?
    private var readCharacteristic: CBMutableCharacteristic?
    private var pairingCharacteristic: CBMutableCharacteristic?

    private var isAdvertising = false
    private var currentNickname: String?
    private var currentArtData: String?

    // コールバック
    var onDataReceived: ((String, CBCentral) -> Void)?
    var onDeviceConnected: ((CBCentral) -> Void)?
    var onDeviceDisconnected: ((CBCentral) -> Void)?
    var onError: ((String) -> Void)?
    var onAdvertisingStarted: (() -> Void)?
    var onAdvertisingStopped: (() -> Void)?

    // 接続中のデバイス
    private var connectedCentrals = Set<CBCentral>()

    // MARK: - Initialization

    override init() {
        super.init()
        peripheralManager = CBPeripheralManager(delegate: self, queue: nil)
    }

    // MARK: - Public Methods

    /// アドバタイズ開始
    func startAdvertising(nickname: String, artData: String) -> Bool {
        guard peripheralManager.state == .poweredOn else {
            print("[BlePeripheralManager] Bluetooth is not powered on")
            onError?("Bluetoothがオンになっていません")
            return false
        }

        if isAdvertising {
            print("[BlePeripheralManager] Already advertising")
            return true
        }

        currentNickname = nickname
        currentArtData = artData

        // サービスセットアップ
        setupService()

        // アドバタイズデータ
        let advertisementData: [String: Any] = [
            CBAdvertisementDataServiceUUIDsKey: [BlePeripheralManager.serviceUUID],
            CBAdvertisementDataLocalNameKey: "\(BlePeripheralManager.advertisePrefix)\(nickname)"
        ]

        // アドバタイズ開始
        peripheralManager.startAdvertising(advertisementData)
        print("[BlePeripheralManager] Advertising started with nickname: \(nickname)")

        return true
    }

    /// アドバタイズ停止
    func stopAdvertising() {
        if !isAdvertising {
            return
        }

        peripheralManager.stopAdvertising()

        // サービス削除
        if let service = service {
            peripheralManager.remove(service)
        }

        service = nil
        writeCharacteristic = nil
        readCharacteristic = nil
        pairingCharacteristic = nil
        connectedCentrals.removeAll()

        isAdvertising = false
        print("[BlePeripheralManager] Advertising stopped")

        onAdvertisingStopped?()
    }

    /// サービスセットアップ
    private func setupService() {
        // サービス作成
        let service = CBMutableService(type: BlePeripheralManager.serviceUUID, primary: true)

        // Write Characteristic
        writeCharacteristic = CBMutableCharacteristic(
            type: BlePeripheralManager.writeCharUUID,
            properties: [.write, .writeWithoutResponse],
            value: nil,
            permissions: .writeable
        )

        // Read Characteristic (Notify)
        readCharacteristic = CBMutableCharacteristic(
            type: BlePeripheralManager.readCharUUID,
            properties: [.read, .notify],
            value: nil,
            permissions: .readable
        )

        // Pairing Characteristic
        pairingCharacteristic = CBMutableCharacteristic(
            type: BlePeripheralManager.pairingCharUUID,
            properties: [.read, .write],
            value: nil,
            permissions: [.readable, .writeable]
        )

        // キャラクタリスティック追加
        service.characteristics = [
            writeCharacteristic!,
            readCharacteristic!,
            pairingCharacteristic!
        ]

        // サービス追加
        peripheralManager.add(service)
        self.service = service

        print("[BlePeripheralManager] Service setup completed")
    }

    /// Centralにデータ送信（Notify）
    private func sendDataToCentral(_ central: CBCentral) {
        guard let artData = currentArtData else {
            return
        }

        guard let characteristic = readCharacteristic else {
            return
        }

        guard let data = artData.data(using: .utf8) else {
            return
        }

        let success = peripheralManager.updateValue(
            data,
            for: characteristic,
            onSubscribedCentrals: [central]
        )

        if success {
            print("[BlePeripheralManager] Notified data to central: \(data.count) bytes")
        } else {
            print("[BlePeripheralManager] Failed to notify data (queue full)")
        }
    }

    /// アドバタイズ状態確認
    func getIsAdvertising() -> Bool {
        return isAdvertising
    }

    /// 接続デバイス数
    func getConnectedDeviceCount() -> Int {
        return connectedCentrals.count
    }

    /// クリーンアップ
    func cleanup() {
        stopAdvertising()
        currentNickname = nil
        currentArtData = nil
    }
}

// MARK: - CBPeripheralManagerDelegate

extension BlePeripheralManager: CBPeripheralManagerDelegate {
    func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
        switch peripheral.state {
        case .poweredOn:
            print("[BlePeripheralManager] Bluetooth powered on")
        case .poweredOff:
            print("[BlePeripheralManager] Bluetooth powered off")
            if isAdvertising {
                isAdvertising = false
                onAdvertisingStopped?()
            }
        case .resetting:
            print("[BlePeripheralManager] Bluetooth resetting")
        case .unauthorized:
            print("[BlePeripheralManager] Bluetooth unauthorized")
            onError?("Bluetooth権限がありません")
        case .unsupported:
            print("[BlePeripheralManager] Bluetooth not supported")
            onError?("このデバイスはBluetoothをサポートしていません")
        case .unknown:
            print("[BlePeripheralManager] Bluetooth state unknown")
        @unknown default:
            print("[BlePeripheralManager] Bluetooth unknown state")
        }
    }

    func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: Error?) {
        if let error = error {
            print("[BlePeripheralManager] Advertising failed: \(error.localizedDescription)")
            onError?("アドバタイズ失敗: \(error.localizedDescription)")
            isAdvertising = false
        } else {
            print("[BlePeripheralManager] Advertising started successfully")
            isAdvertising = true
            onAdvertisingStarted?()
        }
    }

    func peripheralManager(_ peripheral: CBPeripheralManager, didAdd service: CBService, error: Error?) {
        if let error = error {
            print("[BlePeripheralManager] Failed to add service: \(error.localizedDescription)")
            onError?("サービス追加失敗: \(error.localizedDescription)")
        } else {
            print("[BlePeripheralManager] Service added successfully")
        }
    }

    func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
        for request in requests {
            if request.characteristic.uuid == BlePeripheralManager.writeCharUUID {
                // データ受信
                if let value = request.value, let receivedData = String(data: value, encoding: .utf8) {
                    print("[BlePeripheralManager] Received data: \(receivedData.prefix(50))...")

                    // 受信データをコールバック
                    onDataReceived?(receivedData, request.central)

                    // Note: データ送信はdidSubscribeToで既に実行済みのため、ここでは送信しない
                    // これにより二重送信を防ぐ
                }

                // レスポンス送信
                peripheral.respond(to: request, withResult: .success)
            } else {
                print("[BlePeripheralManager] Unknown characteristic write: \(request.characteristic.uuid)")
                peripheral.respond(to: request, withResult: .requestNotSupported)
            }
        }
    }

    func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveRead request: CBATTRequest) {
        if request.characteristic.uuid == BlePeripheralManager.readCharUUID {
            // 自分のデータを返す
            if let artData = currentArtData, let data = artData.data(using: .utf8) {
                if request.offset > data.count {
                    peripheral.respond(to: request, withResult: .invalidOffset)
                    return
                }

                let range = request.offset..<data.count
                request.value = data.subdata(in: range)

                print("[BlePeripheralManager] Read request, sending \(request.value?.count ?? 0) bytes")
                peripheral.respond(to: request, withResult: .success)
            } else {
                peripheral.respond(to: request, withResult: .unlikelyError)
            }
        } else {
            print("[BlePeripheralManager] Unknown characteristic read: \(request.characteristic.uuid)")
            peripheral.respond(to: request, withResult: .requestNotSupported)
        }
    }

    func peripheralManager(_ peripheral: CBPeripheralManager, central: CBCentral, didSubscribeTo characteristic: CBCharacteristic) {
        print("[BlePeripheralManager] Central subscribed to characteristic: \(characteristic.uuid)")
        connectedCentrals.insert(central)
        onDeviceConnected?(central)
        
        // Readキャラクタリスティックに購読した時、自動的にデータを送信
        // これにより、Central側がNotify購読後すぐにデータを受信できる
        if characteristic.uuid == BlePeripheralManager.readCharUUID {
            print("[BlePeripheralManager] Central subscribed to read characteristic, sending data")
            sendDataToCentral(central)
        }
    }

    func peripheralManager(_ peripheral: CBPeripheralManager, central: CBCentral, didUnsubscribeFrom characteristic: CBCharacteristic) {
        print("[BlePeripheralManager] Central unsubscribed from characteristic: \(characteristic.uuid)")
        connectedCentrals.remove(central)
        onDeviceDisconnected?(central)
    }
}
